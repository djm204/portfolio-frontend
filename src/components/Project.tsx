import React from 'react';
import cnames from 'classnames';

type ProjectProps = {
    name: string,
    points: Array<string> | string,
    role: string,
    className: string,
    dataAos?: string,
    dataAosDuration?: string,
    dataAosEasing?: string,
    dataAosOnce?: string
}

const Project = (props: ProjectProps) => {
    const { role, name, points, className, dataAos, dataAosDuration, dataAosEasing, dataAosOnce } = props;

    return (
        <div className={cnames("Project", className)}
            data-aos={dataAos}
            data-aos-duration={dataAosDuration}
            data-aos-easing={dataAosEasing}
            data-aos-once={dataAosOnce}
        >
            <h3>{name}</h3>
            <div>{role}</div>
            <ul>
                {typeof points !== 'string' ?
                    points.map(point => <li key={point}>{point}</li>) :
                    <li>{points}</li>}
            </ul>
            <div className="Project__Picture"></div>
        </div>
    );
}

export default Project;
