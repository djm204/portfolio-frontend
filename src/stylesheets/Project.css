.Project {
  padding: 5px;
  position: relative;
  letter-spacing: 0.05em;
  display: flex;
  align-items: center;
  justify-content: center;
  max-width: 1200px;
  width: 100%;
}

.Project--hidden {
  visibility: hidden;
  opacity: 0;
  transition: all 0.2s ease-in-out;
}

.Project--visible {
  visibility: visible;
  opacity: 1;
  transition: all 1s ease-in-out;
}

.Project--SlideIn {
  visibility: visible;
  transform: translateX(150px);
  animation: slideLeft 0.8s ease forwards;
}

.Project--SlideIn:nth-child(odd) {
  animation-duration: 0.6s;
}

.Project__Role {
  font-style: italic;
  transition: all 1s ease-in-out;
}

.Project__Details {
  transition: all 1s ease-in-out;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  height: 250px;
  flex-basis: 60%;
  margin-left: 20px;
  background: rgba(172, 197, 255, 0.4);
  padding: 20px;
  box-sizing: initial;
}

.Project__Image {
  width: 100%;
  height: 100%;
  flex-basis: 50%;
  margin-right: 10px;
  object-fit: contain;
  padding-right: 20px;
  max-height: 275px;
  cursor: pointer;
}

.ExpandedProjectImageBackground {
  flex-direction: column;
  justify-content: center;
  align-items: center;
  display: none;
}

.ExpandedProjectImageBackground.--expanded {
  display: flex;
  position: fixed;
  top: 0;
  left: 0;
  z-index: 2;
  width: 100vw;
  height: 100vh;
  text-align: center;
  background-color: rgba(0, 0, 0, 0.4);
  transform: scaleY(1);
}

.Project__Image__Wrapper__Expanded {
  width: 75%;
}

.Project__Image__Expanded {
  padding: 0;
  margin: 0;
  width: 75%;
  flex-basis: 100%;
}

.Project__Notables {
  list-style: none;
}

.Project__NotPublic {
  font-weight: bold;
}

.Project__Link__Wrapper {
  flex-grow: 1;
  display: flex;
  justify-content: flex-end;
}
.Project__Link,
.Project__Link a {
  color: rgb(0, 34, 111);
}

.Project__Link {
  display: inline-block;
  padding: 10px 20px;
  border-radius: 3px;
  border: 1px solid rgba(26, 95, 255, 1);
  margin-top: auto;
}

.Project__Link:hover {
  text-decoration: none;
  color: rgba(172, 197, 255, 1);
  background-color: rgba(26, 95, 255, 1);
  border: 1px solid rgb(0, 34, 111);
  transition: background-color ease-in-out 0.2s;
}
