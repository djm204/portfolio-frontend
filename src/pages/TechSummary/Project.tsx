import React, { useState } from 'react'
import cnames from 'classnames'
import '../../stylesheets/Project.css'

type ProjectProps = {
  name: string
  link: string
  imgSrc: string
  points: Array<string> | string
  role: string
  className: string
  dataAos?: string
  dataAosDuration?: string
  dataAosEasing?: string
  dataAosOnce?: string
}

const Project = (props: ProjectProps) => {
  const [isExpanded, setIsExpanded] = useState(false)
  const { role, name, link, points, className, imgSrc, dataAosDuration, dataAosEasing } = props

  const handleClick = (event: React.MouseEvent<HTMLImageElement, MouseEvent>) => {
    if (event) event.persist()
    setIsExpanded(!isExpanded)
  }

  return (
    <div id={className} className={cnames('Project', className)}>
      <div
        onClick={handleClick}
        className={cnames('ExpandedProjectImageBackground', { '--expanded': isExpanded })}
      >
        <div className="Project__Image__Wrapper__Expanded">
          <img className="Project__Image__Expanded" alt={name} src={imgSrc} />
        </div>
      </div>
      <img onClick={handleClick} className={cnames('Project__Image')} alt={name} src={imgSrc} />

      <div className={cnames('Project__Details', { '--expanded': isExpanded })}>
        <h3
          className="Project__Name "
          data-aos-duration={dataAosDuration}
          data-aos-easing={dataAosEasing}
        >
          {name}
        </h3>
        <div className="Project__Role ">{role}</div>
        <ul className="Project__Notables ">
          {typeof points !== 'string' ? (
            points.map(point => <li key={point}>{point}</li>)
          ) : (
            <li>{points}</li>
          )}
        </ul>
        <div className="Project__Link__Wrapper">
          {link === 'NPA' ? (
            <span className="Project__NotPublic ">*Not Publicly Available</span>
          ) : (
            <a href={link} rel="noopener noreferrer" target="_blank" className="Project__Link">
              View Project
            </a>
          )}
        </div>
      </div>
    </div>
  )
}

export default Project
