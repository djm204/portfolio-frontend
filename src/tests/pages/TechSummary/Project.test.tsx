import React from 'react'
import Project from '../../../pages/TechSummary/Project'
import { render, fireEvent } from '@testing-library/react'

it('Project renders without crashing', () => {
  const result = render(
    <Project
      name="GoLang"
      link="https://someproject.com"
      imgExpandedSrc="imgSrcExp"
      imgSrc="imgSrc"
      role="bestRole"
      className="classy"
      points={[]}
    />
  )

  expect(result).toMatchSnapshot()
})

it('Expanded Image renders image when Project Image is clicked', () => {
  const result = render(
    <Project
      name="GoLang"
      link="https://someproject.com"
      imgExpandedSrc="imgSrcExp"
      imgSrc="imgSrc"
      role="bestRole"
      className="classy"
      points={[]}
    />
  )
  const project = result.container
  const projectImage = project.querySelector('.Project__Image')

  fireEvent.click(projectImage as Element | Document | Window)

  expect(project.querySelectorAll('.ExpandedProjectImageBackground.--expanded')).toHaveLength(1)
})
