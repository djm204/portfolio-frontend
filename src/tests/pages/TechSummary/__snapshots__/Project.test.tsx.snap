// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Project classy"
        id="classy"
      >
        <div
          class="ExpandedProjectImageBackground"
        >
          <div
            class="Project__Image__Wrapper__Expanded"
          >
            <img
              alt="GoLang"
              class="Project__Image__Expanded"
              src="imgSrcExp"
            />
          </div>
        </div>
        <img
          alt="GoLang"
          class="Project__Image"
          src="imgSrc"
        />
        <div
          class="Project__Details"
        >
          <h3
            class="Project__Name "
          >
            GoLang
          </h3>
          <div
            class="Project__Role "
          >
            bestRole
          </div>
          <ul
            class="Project__Notables "
          />
          <div
            class="Project__Link__Wrapper"
          >
            <a
              class="Project__Link"
              href="https://someproject.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              View Project
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Project classy"
      id="classy"
    >
      <div
        class="ExpandedProjectImageBackground"
      >
        <div
          class="Project__Image__Wrapper__Expanded"
        >
          <img
            alt="GoLang"
            class="Project__Image__Expanded"
            src="imgSrcExp"
          />
        </div>
      </div>
      <img
        alt="GoLang"
        class="Project__Image"
        src="imgSrc"
      />
      <div
        class="Project__Details"
      >
        <h3
          class="Project__Name "
        >
          GoLang
        </h3>
        <div
          class="Project__Role "
        >
          bestRole
        </div>
        <ul
          class="Project__Notables "
        />
        <div
          class="Project__Link__Wrapper"
        >
          <a
            class="Project__Link"
            href="https://someproject.com"
            rel="noopener noreferrer"
            target="_blank"
          >
            View Project
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
